File: meshgrid.py
"""
Meshgrid Module

This module provides meshgrid functionality for 2D and 3D grids.
Ported from Fortran to Python.

Author: Victor Carreira (Original Fortran)
Python Port: 2026
"""

import numpy as np
from typing import Tuple


def meshgrid2d(x: np.ndarray, y: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
    """
    Create a 2D meshgrid from 1D arrays.
    
    This is a wrapper around numpy.meshgrid for compatibility with the
    original Fortran implementation.
    
    Parameters
    ----------
    x : np.ndarray
        1D array of x coordinates
    y : np.ndarray
        1D array of y coordinates
    
    Returns
    -------
    X, Y : np.ndarray
        2D arrays representing the grid coordinates
    
    Examples
    --------
    >>> x = np.linspace(0, 1, 5)
    >>> y = np.linspace(0, 1, 5)
    >>> X, Y = meshgrid2d(x, y)
    >>> X.shape
    (5, 5)
    """
    X, Y = np.meshgrid(x, y, indexing='ij')
    return X, Y


def meshgrid3d(x: np.ndarray, y: np.ndarray, z: np.ndarray) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
    """
    Create a 3D meshgrid from 1D arrays.
    
    This is a wrapper around numpy.meshgrid for compatibility with the
    original Fortran implementation.
    
    Parameters
    ----------
    x : np.ndarray
        1D array of x coordinates
    y : np.ndarray
        1D array of y coordinates
    z : np.ndarray
        1D array of z coordinates
    
    Returns
    -------
    X, Y, Z : np.ndarray
        3D arrays representing the grid coordinates
    
    Examples
    --------
    >>> x = np.linspace(0, 1, 3)
    >>> y = np.linspace(0, 1, 3)
    >>> z = np.linspace(0, 1, 3)
    >>> X, Y, Z = meshgrid3d(x, y, z)
    >>> X.shape
    (3, 3, 3)
    """
    X, Y, Z = np.meshgrid(x, y, z, indexing='ij')
    return X, Y, Z


def create_linear_grid(start: float, end: float, num_points: int) -> np.ndarray:
    """
    Create a linearly spaced 1D array.
    
    Parameters
    ----------
    start : float
        Starting value
    end : float
        Ending value
    num_points : int
        Number of points
    
    Returns
    -------
    np.ndarray
        Linearly spaced array
    
    Examples
    --------
    >>> grid = create_linear_grid(0, 10, 5)
    >>> len(grid)
    5
    """
    return np.linspace(start, end, num_points)


if __name__ == '__main__':
    # Example usage matching the original Fortran program
    nx, ny, nz = 50, 50, 50
    
    # Create 1D arrays
    x = create_linear_grid(0.0, 1.0, nx)
    y = create_linear_grid(0.0, 1.0, ny)
    z = create_linear_grid(0.0, 1.0, nz)
    
    print(f'Shape of x: {x.shape}')
    print(f'Shape of y: {y.shape}')
    print(f'Shape of z: {z.shape}')
    
    # Create 2D meshgrid
    X2d, Y2d = meshgrid2d(x, y)
    print(f'\n2D Meshgrid:')
    print(f'Shape of X2d: {X2d.shape}')
    print(f'Shape of Y2d: {Y2d.shape}')
    
    # Create 3D meshgrid
    X3d, Y3d, Z3d = meshgrid3d(x, y, z)
    print(f'\n3D Meshgrid:')
    print(f'Shape of X3d: {X3d.shape}')
    print(f'Shape of Y3d: {Y3d.shape}')
    print(f'Shape of Z3d: {Z3d.shape}')
    
    # Test reshaping (similar to Fortran reshape test)
    x_test = np.linspace(0, 1, 50)
    print(f'\nOriginal shape: {x_test.shape}')
    x_reshaped = x_test.reshape(25, 2)
    print(f'Reshaped to (25, 2): {x_reshaped.shape}')

File: src/meshgrid.rs
//! Meshgrid Module
//!
//! This module provides meshgrid functionality for 2D and 3D grids.
//! Ported from Fortran to Rust.
//!
//! Author: Victor Carreira (Original Fortran)
//! Rust Port: 2026

/// Create a linearly spaced 1D array.
///
/// # Arguments
///
/// * `start` - Starting value
/// * `end` - Ending value
/// * `num_points` - Number of points
///
/// # Returns
///
/// Vector of linearly spaced values
pub fn linspace(start: f64, end: f64, num_points: usize) -> Vec<f64> {
    if num_points == 0 {
        return Vec::new();
    }
    
    if num_points == 1 {
        return vec![start];
    }

    let step = (end - start) / (num_points - 1) as f64;
    (0..num_points)
        .map(|i| start + step * i as f64)
        .collect()
}

/// Create a 2D meshgrid from 1D arrays.
///
/// # Arguments
///
/// * `x` - 1D array of x coordinates
/// * `y` - 1D array of y coordinates
///
/// # Returns
///
/// Tuple of (X, Y) - 2D arrays representing the grid coordinates
/// Each grid is represented as a vector of vectors (rows)
pub fn meshgrid2d(x: &[f64], y: &[f64]) -> (Vec<Vec<f64>>, Vec<Vec<f64>>) {
    let nx = x.len();
    let ny = y.len();

    let mut grid_x = Vec::with_capacity(nx);
    let mut grid_y = Vec::with_capacity(nx);

    for &xi in x.iter() {
        grid_x.push(vec![xi; ny]);
    }

    for _ in 0..nx {
        grid_y.push(y.to_vec());
    }

    (grid_x, grid_y)
}

/// Create a 3D meshgrid from 1D arrays.
///
/// # Arguments
///
/// * `x` - 1D array of x coordinates
/// * `y` - 1D array of y coordinates
/// * `z` - 1D array of z coordinates
///
/// # Returns
///
/// Tuple of (X, Y, Z) - 3D arrays representing the grid coordinates
/// Each grid is represented as a vector of vector of vectors
pub fn meshgrid3d(
    x: &[f64],
    y: &[f64],
    z: &[f64],
) -> (Vec<Vec<Vec<f64>>>, Vec<Vec<Vec<f64>>>, Vec<Vec<Vec<f64>>>) {
    let nx = x.len();
    let ny = y.len();
    let nz = z.len();

    let mut grid_x = Vec::with_capacity(nx);
    let mut grid_y = Vec::with_capacity(nx);
    let mut grid_z = Vec::with_capacity(nx);

    for &xi in x.iter() {
        let mut slice_x = Vec::with_capacity(ny);
        let mut slice_y = Vec::with_capacity(ny);
        let mut slice_z = Vec::with_capacity(ny);

        for &yi in y.iter() {
            slice_x.push(vec![xi; nz]);
            slice_y.push(vec![yi; nz]);
            slice_z.push(z.to_vec());
        }

        grid_x.push(slice_x);
        grid_y.push(slice_y);
        grid_z.push(slice_z);
    }

    (grid_x, grid_y, grid_z)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_linspace() {
        let result = linspace(0.0, 1.0, 5);
        assert_eq!(result.len(), 5);
        assert_eq!(result[0], 0.0);
        assert_eq!(result[4], 1.0);
        assert!((result[2] - 0.5).abs() < 1e-10);
    }

    #[test]
    fn test_meshgrid2d() {
        let x = vec![1.0, 2.0, 3.0];
        let y = vec![4.0, 5.0];

        let (grid_x, grid_y) = meshgrid2d(&x, &y);

        assert_eq!(grid_x.len(), 3);
        assert_eq!(grid_x[0].len(), 2);
        assert_eq!(grid_x[0][0], 1.0);
        assert_eq!(grid_x[2][1], 3.0);

        assert_eq!(grid_y.len(), 3);
        assert_eq!(grid_y[0].len(), 2);
        assert_eq!(grid_y[0][0], 4.0);
        assert_eq!(grid_y[2][1], 5.0);
    }

    #[test]
    fn test_meshgrid3d() {
        let x = vec![1.0, 2.0];
        let y = vec![3.0, 4.0];
        let z = vec![5.0, 6.0];

        let (grid_x, grid_y, grid_z) = meshgrid3d(&x, &y, &z);

        assert_eq!(grid_x.len(), 2);
        assert_eq!(grid_x[0].len(), 2);
        assert_eq!(grid_x[0][0].len(), 2);

        assert_eq!(grid_x[0][0][0], 1.0);
        assert_eq!(grid_y[0][0][0], 3.0);
        assert_eq!(grid_z[0][0][0], 5.0);
    }
}
